### ISSUE TITLE
Runtime Error after modifying binary_mnist.py

### ISSUE THREAD

**Issue by zamyers** [hpi-xnor/BMXNet#10]
For bugs or installation issues, please provide the following information.
The more information you provide, the more likely people will be able to help you.

## Environment info
Operating System:
macOS
Compiler:

Package used (Python/R/Scala/Julia):
Python
MXNet version:

Or if installed from source:

MXNet commit hash (`git rev-parse HEAD`):

If you are using python package, please provide

Python version and distribution:

If you are using R package, please provide

R `sessionInfo()`:

## Error Message:
Please paste the full error message, including stack trace.

[16:20:03] /Users/zamyers/mxnet/mxnet/nnvm/include/dmlc/logging.h:300: [16:20:03] /Users/zamyers/mxnet/mxnet/smd_hpi/src/./q_convolution-inl.h:149: Check failed: data.shape_[1] % mxnet::op::xnor_cpu::BITS_PER_BINARY_WORD == 0 (1 vs. 0) input channel currently have to be multiple of 64 but are: 1

Stack trace returned 18 entries:
[bt] (0) 0   libmxnet.dylib                      0x000000010e275511 _ZN4dmlc15LogMessageFatalD2Ev + 49
[bt] (1) 1   libmxnet.dylib                      0x000000010e2639b5 _ZN4dmlc15LogMessageFatalD1Ev + 21
[bt] (2) 2   libmxnet.dylib                      0x000000010fcd8ae7 _ZN5mxnet2op14QConvolutionOpIN7mshadow3cpuEfE7ForwardERKNS_9OpContextERKNSt3__16vectorINS_5TBlobENS8_9allocatorISA_EEEERKNS9_INS_9OpReqTypeENSB_ISG_EEEESF_SF_ + 2951
[bt] (3) 3   libmxnet.dylib                      0x000000010e391304 _ZN5mxnet4exec17ForwardOpExecutor3RunENS_10RunContextE + 116
[bt] (4) 4   libmxnet.dylib                      0x000000010e3de1e9 _ZZN5mxnet4exec13GraphExecutor13InitCachedOpsEvENK3$_3clENS_10RunContextENS_6engine18CallbackOnCompleteE + 185
[bt] (5) 5   libmxnet.dylib                      0x000000010e3de11f _ZNSt3__128__invoke_void_return_wrapperIvE6__callIJRZN5mxnet4exec13GraphExecutor13InitCachedOpsEvE3$_3NS3_10RunContextENS3_6engine18CallbackOnCompleteEEEEvDpOT_ + 175
[bt] (6) 6   libmxnet.dylib                      0x000000010e3dde99 _ZNSt3__110__function6__funcIZN5mxnet4exec13GraphExecutor13InitCachedOpsEvE3$_3NS_9allocatorIS5_EEFvNS2_10RunContextENS2_6engine18CallbackOnCompleteEEEclEOS8_OSA_ + 73
[bt] (7) 7   libmxnet.dylib                      0x000000010e346aa3 _ZNKSt3__18functionIFvN5mxnet10RunContextENS1_6engine18CallbackOnCompleteEEEclES2_S4_ + 179
[bt] (8) 8   libmxnet.dylib                      0x000000010e37187f _ZN5mxnet6engine14ThreadedEngine15ExecuteOprBlockENS_10RunContextEPNS0_8OprBlockE + 1199
[bt] (9) 9   libmxnet.dylib                      0x000000010e37aa0e _ZN5mxnet6engine23ThreadedEnginePerDevice9CPUWorkerILN4dmlc19ConcurrentQueueTypeE0EEEvPNS1_17ThreadWorkerBlockIXT_EEE + 94
[bt] (10) 10  libmxnet.dylib                      0x000000010e37a99f _ZZZN5mxnet6engine23ThreadedEnginePerDevice13PushToExecuteEPNS0_8OprBlockEbENKUlvE_clEvENKUlvE_clEv + 31
[bt] (11) 11  libmxnet.dylib                      0x000000010e37a96d _ZNSt3__128__invoke_void_return_wrapperIvE6__callIJRZZN5mxnet6engine23ThreadedEnginePerDevice13PushToExecuteEPNS4_8OprBlockEbENKUlvE_clEvEUlvE_EEEvDpOT_ + 45
[bt] (12) 12  libmxnet.dylib                      0x000000010e37a889 _ZNSt3__110__function6__funcIZZN5mxnet6engine23ThreadedEnginePerDevice13PushToExecuteEPNS3_8OprBlockEbENKUlvE_clEvEUlvE_NS_9allocatorIS8_EEFvvEEclEv + 41
[bt] (13) 13  libmxnet.dylib                      0x000000010e373dae _ZNKSt3__18functionIFvvEEclEv + 126
[bt] (14) 14  libmxnet.dylib                      0x000000010e3738f2 _ZNSt3__114__thread_proxyINS_5tupleIJNS_8functionIFvvEEEEEEEEPvS6_ + 402
[bt] (15) 15  libsystem_pthread.dylib             0x00007fffcad9793b _pthread_body + 180
[bt] (16) 16  libsystem_pthread.dylib             0x00007fffcad97887 _pthread_body + 0
[bt] (17) 17  libsystem_pthread.dylib             0x00007fffcad9708d thread_start + 13

[16:20:03] /Users/zamyers/mxnet/mxnet/nnvm/include/dmlc/logging.h:300: [16:20:03] /Users/zamyers/mxnet/mxnet/src/engine/./threaded_engine.h:336: [16:20:03] /Users/zamyers/mxnet/mxnet/smd_hpi/src/./q_convolution-inl.h:149: Check failed: data.shape_[1] % mxnet::op::xnor_cpu::BITS_PER_BINARY_WORD == 0 (1 vs. 0) input channel currently have to be multiple of 64 but are: 1

## Minimum reproducible example
if you are using your own code, please provide a short script that reproduces the error.

Replaced 

"""
def get_binary_lenet():
	data = mx.symbol.Variable('data')

	# first conv layer
	conv1 = mx.sym.Convolution(data=data, kernel=(5,5), num_filter=64)	
	tanh1 = mx.sym.Activation(data=conv1, act_type="tanh")
	pool1 = mx.sym.Pooling(data=tanh1, pool_type="max", kernel=(2,2), stride=(2,2))
	bn1 = mx.sym.BatchNorm(data=pool1)

	# second conv layer
	ba1 = mx.sym.QActivation(data=bn1, act_bit=BITA, backward_only=True)
	conv2 = mx.sym.QConvolution(data=ba1, kernel=(5,5), num_filter=64, act_bit=BITW)
	bn2 = mx.sym.BatchNorm(data=conv2)
	pool2 = mx.sym.Pooling(data=bn2, pool_type="max", kernel=(2,2), stride=(2,2))
	
	# first fullc layer
	flatten = mx.sym.Flatten(data=pool2)	
	ba2 = mx.sym.QActivation(data=flatten,  act_bit=BITA, backward_only=True)	
	fc1 = mx.symbol.QFullyConnected(data=ba2, num_hidden=1000, act_bit=BITW)
	#fc1 = mx.sym.Custom(data=fc1, op_type='debug')
	bn3 = mx.sym.BatchNorm(data=fc1)
	tanh3 = mx.sym.Activation(data=bn3, act_type="tanh")

	# second fullc
	fc2 = mx.sym.FullyConnected(data=tanh3, num_hidden=10)
	# softmax loss
	lenet = mx.sym.SoftmaxOutput(data=fc2, name='softmax')

	print 'using quantized lenet with bitwidth %d (weights), %d (activations) and %d (gradients)' % (BITW, BITA, BITG)
	return lenet
"""

with

"""
def get_binary_lenet():
	data = mx.symbol.Variable('data')

	# first conv layer
	ba0 = mx.sym.QActivation(data=data, act_bit=BITA, backward_only=True)
	conv0 = mx.sym.QConvolution(data=ba0, kernel=(5,5), num_filter=64, act_bit=BITW)
	bn0 = mx.sym.BatchNorm(data=conv0)
	pool0 = mx.sym.Pooling(data=bn0, pool_type="max", kernel=(2,2), stride=(2,2))

	# second conv layer
	ba1 = mx.sym.QActivation(data=pool0, act_bit=BITA, backward_only=True)
	conv2 = mx.sym.QConvolution(data=ba1, kernel=(5,5), num_filter=64, act_bit=BITW)
	bn2 = mx.sym.BatchNorm(data=conv2)
	pool2 = mx.sym.Pooling(data=bn2, pool_type="max", kernel=(2,2), stride=(2,2))
	
	# first fullc layer
	flatten = mx.sym.Flatten(data=pool2)	
	ba2 = mx.sym.QActivation(data=flatten,  act_bit=BITA, backward_only=True)	
	fc1 = mx.symbol.QFullyConnected(data=ba2, num_hidden=1000, act_bit=BITW)
	#fc1 = mx.sym.Custom(data=fc1, op_type='debug')
	bn3 = mx.sym.BatchNorm(data=fc1)
	tanh3 = mx.sym.Activation(data=bn3, act_type="tanh")

	# second fullc
	fc2 = mx.sym.QFullyConnected(data=tanh3, num_hidden=10, act_bit=BITW)
	# softmax loss
	lenet = mx.sym.SoftmaxOutput(data=fc2, name='softmax')

	print 'using quantized lenet with bitwidth %d (weights), %d (activations) and %d (gradients)' % (BITW, BITA, BITG)
	return lenet
"""

## Steps to reproduce
or if you are running standard examples, please provide the commands you have run that lead to the error.

1. Quantize the first layer of the network by duplicating the middle binary convolutional layer (code is provided above)
2. Ran BMXnet without CUDA/GPU enabled
3. Error occurs

## What have you tried to solve it?

1. Replacing Qconvolution with a normal convolution eliminates the error but the network no longer trains. Even with quantization activation set to 32b.

**Comment by yanghaojin** [comment]
Hi, this issue is due to the input channel number of the first conv layer can not be divided by 32 or 64 which depends on you binary word setting. for binarization we will later concatinate 32 or 64 float weight parameters into one integer parameter, because of that the input channel currently have to be multiple of 32 or 64. On the other hand you'd better not binarize the first layer and last layer since it will significantly decrease the accuracy according to the previous experimental results.
